File: RIRImageResizing/Classes/NSObject/RIRResizeImageOperation.m
//
//  RIRResizeImageOperation.m
//  Pods
//
//  Created by Richard Reitzfeld on 11/21/16.
//
//

#import "RIRResizeImageOperation.h"
#import "RIRResizeImageOperationParameters.h"

#import <ResplendentUtilities/RUConditionalReturn.h>





@interface RIRResizeImageOperation ()

#pragma mark - scale
/**
 Returns a new `UIImage` instance, which is a resized version of the receiver.
 
 @param newSize    The size that the new UIImage instance aims for. This is not necessarily the size of the final image, as it depends on the target image's size, and the `resizeMode` parameter.
 @param resizeMode Indicates which type of resize method to use.
 
 @return A new `UIImage` instance, which is a resized version of the receiver.
 */
-(nullable UIImage*)scaleToSize:(CGSize)newSize usingMode:(UIImage_RIRResizing_ResizeMode)resizeMode;

/**
 Returns a new `UIImage` instance, which is a resized version of the receiver. Uses `UIImage_RIRResizing_ResizeMode_aspectFit` resizing mode.
 
 @param newSize    The size that the new UIImage instance aims for. This is not necessarily the size of the final image, as it depends on the target image's size, and the `resizeMode` parameter.
 
 @return A new `UIImage` instance, which is a resized version of the receiver.
 */
-(nonnull UIImage*)scaleToFitSize:(CGSize)newSize;		//UIImage_RIRResizing_ResizeMode_aspectFit

/**
 Returns a new `UIImage` instance, which is a resized version of the receiver. Uses `UIImage_RIRResizing_ResizeMode_scaleToFill` resizing mode.
 
 @param newSize    The size that the new UIImage instance aims for. This is not necessarily the size of the final image, as it depends on the target image's size, and the `resizeMode` parameter.
 
 @return A new `UIImage` instance, which is a resized version of the receiver.
 */
-(nonnull UIImage*)scaleToFillSize:(CGSize)newSize;		//UIImage_RIRResizing_ResizeMode_scaleToFill

/**
 Returns a new `UIImage` instance, which is a resized version of the receiver. Uses `UIImage_RIRResizing_ResizeMode_aspectFill` resizing mode.
 
 @param newSize    The size that the new UIImage instance aims for. This is not necessarily the size of the final image, as it depends on the target image's size, and the `resizeMode` parameter.
 
 @return A new `UIImage` instance, which is a resized version of the receiver.
 */
-(nonnull UIImage*)scaleToCoverSize:(CGSize)newSize;	//UIImage_RIRResizing_ResizeMode_aspectFill

#pragma mark - resizeParameters
@property (nonatomic, strong, nonnull) RIRResizeImageOperationParameters* resizeParameters;

#pragma mark - image
@property (nonatomic, strong, nonnull) UIImage* image;


@end





@implementation RIRResizeImageOperation

#pragma mark - NSObject
-(instancetype)init
{
    kRUConditionalReturn_ReturnValueNil(YES, YES);
    return [self init_with_resizeParameters:[RIRResizeImageOperationParameters new]
                                      image:[UIImage new]];
}

#pragma mark - init
-(instancetype)init_with_resizeParameters:(nonnull RIRResizeImageOperationParameters *)resizeParameters
                                    image:(nonnull UIImage *)image
{
    kRUConditionalReturn_ReturnValueNil(resizeParameters == nil, YES);
    kRUConditionalReturn_ReturnValueNil(image == nil, YES);
    
    if (self = [super init])
    {
        [self setResizeParameters:resizeParameters];
        [self setImage:image];
    }
    
    return self;
}

#pragma mark - scale
-(nullable UIImage*)scaleToSize:(CGSize)newSize usingMode:(UIImage_RIRResizing_ResizeMode)resizeMode
{
    switch (resizeMode)
    {
        case UIImage_RIRResizing_ResizeMode_aspectFit:
            return [self scaleToFitSize:newSize];
            break;
            
        case UIImage_RIRResizing_ResizeMode_aspectFill:
            return [self scaleToCoverSize:newSize];
            break;
            
        case UIImage_RIRResizing_ResizeMode_scaleToFill:
            return [self scaleToFillSize:newSize];
            break;
            
        default:
            break;
    }
    
    NSAssert(false, @"unhandled %li",(long)resizeMode);
    return nil;
}

-(nonnull UIImage*)scaleToFitSize:(CGSize)newSize
{
    /// Keep aspect ratio
    size_t destWidth, destHeight;
    if (self.image.size.width > self.image.size.height)
    {
        destWidth = (size_t)newSize.width;
        destHeight = (size_t)(self.image.size.height * newSize.width / self.image.size.width);
    }
    else
    {
        destHeight = (size_t)newSize.height;
        destWidth = (size_t)(self.image.size.width * newSize.height / self.image.size.height);
    }
    if (destWidth > newSize.width)
    {
        destWidth = (size_t)newSize.width;
        destHeight = (size_t)(self.image.size.height * newSize.width / self.image.size.width);
    }
    if (destHeight > newSize.height)
    {
        destHeight = (size_t)newSize.height;
        destWidth = (size_t)(self.image.size.width * newSize.height / self.image.size.height);
    }
    
    return [self scaleToFillSize:(CGSize){
        .width		= destWidth,
        .height		= destHeight,
    }];
}

-(nonnull UIImage*)scaleToFillSize:(CGSize)newSize
{
    UIGraphicsBeginImageContextWithOptions(newSize, NO, self.resizeParameters.scale);
    
    [self.image drawInRect:CGRectMake(0,0,newSize.width,newSize.height)];
    UIImage* newImage = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return newImage;
}

-(nonnull UIImage*)scaleToCoverSize:(CGSize)newSize
{
    size_t destWidth, destHeight;
    CGFloat const widthRatio = newSize.width / self.image.size.width;
    CGFloat const heightRatio = newSize.height / self.image.size.height;
    /// Keep aspect ratio
    if (heightRatio > widthRatio)
    {
        destHeight = (size_t)newSize.height;
        destWidth = (size_t)(self.image.size.width * newSize.height / self.image.size.height);
    }
    else
    {
        destWidth = (size_t)newSize.width;
        destHeight = (size_t)(self.image.size.height * newSize.width / self.image.size.width);
    }
    
    return [self scaleToFillSize:CGSizeMake(destWidth, destHeight)];
}

#pragma mark - resizedImage
@synthesize resizedImage = _resizedImage;

-(nullable UIImage *)resizedImage
{
    if (_resizedImage == nil)
    {
        _resizedImage = [self scaleToSize:self.resizeParameters.newSize usingMode:self.resizeParameters.resizeMode];
    }
    return _resizedImage;
}

@end

File: RIRImageResizing/Classes/NSObject/RIRResizeImageOperation.swift
//
//  RIRResizeImageOperation.swift
//  Pods-RIRImageResizing_Example
//
//  Created by Benjamin Maer on 4/7/19.
//

import Foundation
import UIKit

public class RIRResizeImageOperation {
    // MARK: - Parameters
    let parameters: RIRResizeImageOperationParameters
    
    // MARK: - Image
    let image: UIImage
    
    // MARK: - Has Attempted Image Resize
    private var hasAttemptedImageResize: Bool = false
    
    // MARK: - Resized Image
    private enum ResizedImagePointer {
        case notSet, attempted, set(UIImage)
    }
    private var resizedImagePointer: ResizedImagePointer = .notSet
    
    enum ResizedImageError: Error {
        case alreadyAttemptedAndFailed
    }
    func resizedImage() throws -> UIImage {
        switch resizedImagePointer {
        case .notSet:
            resizedImagePointer = .attempted
            let resizedImage = try resizedImageGenerate()
            resizedImagePointer = .set(resizedImage)
            return resizedImage
            
        case .attempted:
            throw ResizedImageError.alreadyAttemptedAndFailed
            
        case .set(let image):
            return image
        }
    }
    
    // MARK: - Init
    init(parameters: RIRResizeImageOperationParameters, image: UIImage) {
        self.parameters = parameters
        self.image = image
    }
    
    // MARK: - Resized Image (functions)
    var resizedImageSize: CGSize {
        return {
            switch parameters.resizeType {
            case .scaleToFill:  return resizedImageSizeScaleToFill
            case .aspectFit:    return resizedImageSizeAspectFit
            case .aspectFill:   return resizedImageSizeAspectFill
            }
        }()
    }
    
    private var resizedImageSizeScaleToFill: CGSize {
        return parameters.newSize.bounded(by: image.size)
    }
    
    private var resizedImageSizeAspectFit: CGSize {
        let initialSize = image.size, newSize = parameters.newSize
        var sizeChanges: [(CGSize) -> CGSize] = []
        sizeChanges.append { $0.scaled(to: newSize) }
        sizeChanges.append { $0.boundedWithPreservedScale(by: newSize) }
        sizeChanges.append { $0.boundedWithPreservedScale(by: initialSize) }
        
        var size = initialSize
        sizeChanges.forEach({ sizeFunction in
            size = sizeFunction(size)
        })
        return size
    }
    
    private var resizedImageSizeAspectFill: CGSize {
        let newSize = parameters.newSize
        let widthRatio = newSize.width / image.size.width
        let heightRatio = newSize.height / image.size.height
        let heightRatioIsGreaterThanWidth = heightRatio > widthRatio
        return image.size.boundedWithPreservedScale(by: CGSize(width: heightRatioIsGreaterThanWidth ? image.size.width * heightRatio : newSize.width,
                                                               height: heightRatioIsGreaterThanWidth ? newSize.height : image.size.height * widthRatio))
    }
    
    private func resizedImageGenerate() throws -> UIImage {
        return try resizedImageScaleToFill(size: resizedImageSize)
    }
    
    private func resizedImageScaleToFill() throws -> UIImage {
        return try resizedImageScaleToFill(size: parameters.newSize)
    }
    
    private func resizedImageScaleToFill(size: CGSize) throws -> UIImage {
        return try type(of: self).resized(image: image, scaleToFill: size, scale: scaleToUse)
    }
    
    enum ResizeImageGenerationError: Error {
        case generatedImageWasNil
    }
    private static func resized(image: UIImage, scaleToFill size: CGSize, scale: CGFloat) throws -> UIImage {
        UIGraphicsBeginImageContextWithOptions(size, false, scale);
        image.draw(in: CGRect(origin: .zero, size: size))
        
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        
        UIGraphicsEndImageContext()
        
        guard let finalImage = newImage else {
            throw ResizeImageGenerationError.generatedImageWasNil
        }
        
        return finalImage
    }
    
    // MARK: - Scale
    private var scaleToUse: CGFloat {
        guard
            let scale = parameters.scale,
            scale > 0
            else { return image.scale }
        return scale
    }
}

@objc(RIRResizeImageOperation) public class RIRResizeImageOperationObjCLegacy: NSObject {
    // MARK: - Parameters
    @objc let parameters: RIRResizeImageOperationParametersObjCLegacy
    
    // MARK: - Image
    @objc var image: UIImage { return swiftInstance.image }
    
    // MARK: - Swift Instance
    let swiftInstance: RIRResizeImageOperation
    
    // MARK: - Resized Image
    @objc func resizedImage() throws -> UIImage {
        return try swiftInstance.resizedImage()
    }
    
    // MARK: - Init
    @objc(init_with_parameters:image:) public init(parameters: RIRResizeImageOperationParametersObjCLegacy, image: UIImage) {
        self.parameters = parameters
        swiftInstance = RIRResizeImageOperation(parameters: parameters.swiftInstance, image: image)
    }
}
